### PR TITLE
Make Function submission checks optional

### PR DESCRIPTION
**Pull request by srkukarni**
### Motivation
Currently when users submit functions/source/sinks, the function worker does some pretty extensive checks for java functions. This checks involve loading the user defined classes. Since function-worker is privileged code, but function itself is non-privileged, it could lead to situations where a misconfigured/malignant user-code causes catastrophic cluster failures. 
This pr adds a switch to make these checks optional.
### Modifications

*Describe the modifications you've done.*

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API: (yes / no)
  - The schema: (yes / no / don't know)
  - The default values of configurations: (yes / no)
  - The wire protocol: (yes / no)
  - The rest endpoints: (yes / no)
  - The admin cli options: (yes / no)
  - Anything that affects deployment: (yes / no / don't know)

### Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/source/PulsarSource.java [hunk 1]
```diff
@@ -145,9 +158,42 @@ public void close() throws Exception {
     @VisibleForTesting
     Map<String, ConsumerConfig<T>> setupConsumerConfigs() throws ClassNotFoundException {
         Map<String, ConsumerConfig<T>> configs = new TreeMap<>();
-
-        Class<?> typeArg = Reflections.loadClass(this.pulsarSourceConfig.getTypeClassName(),
-                Thread.currentThread().getContextClassLoader());
+        Class<?> typeArg;
```
Comment (jerrypeng) [inline pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/source/PulsarSource.java]: Is this change related to "Make Function submission checks optional"

Comment (jerrypeng) [reply pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/source/PulsarSource.java]: Can we do these checks in the JavaInstanceRunnable so that we don't complicate the logic here.  If the sourceSpec type classname is not set than and get the types from the jar so we can still get pass the source type.  I guess we also don't need to use a string to represent the type class, you can just pass in Class<?> typeClass in the pulsar source config

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/sink/PulsarSink.java [hunk 2]
```diff
@@ -310,12 +319,28 @@ public void close() throws Exception {
     @SuppressWarnings("unchecked")
     @VisibleForTesting
     Schema<T> initializeSchema() throws ClassNotFoundException {
+        Class<?> typeArg;
         if (StringUtils.isEmpty(this.pulsarSinkConfig.getTypeClassName())) {
-            return (Schema<T>) Schema.BYTES;
+            switch (InstanceUtils.calculateSubjectType(functionDetails)) {
```
Comment (jerrypeng) [inline pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/sink/PulsarSink.java]: Same comment about doing this in JavaInstanceRunnable

**Comment by srkukarni** [comment]
@jerrypeng have made the changes.

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java [hunk 3]
```diff
@@ -66,20 +66,16 @@
...
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import java.io.File;
 import java.io.FileNotFoundException;
-import java.util.Arrays;
-import java.util.Collections;
-import java.util.HashMap;
-import java.util.Map;
+import java.util.*;
```
Comment (jerrypeng) [inline pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java]: can we avoid using wildcards

File: pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionActioner.java [hunk 4]
```diff
@@ -400,7 +399,7 @@ private File getBuiltinArchive(FunctionDetails.Builder functionDetails) throws I
                 builder.setClassName(sourceClass);
                 functionDetails.setSource(builder);
 
-                fillSourceTypeClass(functionDetails, archive, sourceClass);
+                SourceConfigUtils.fillSourceTypeClass(functionDetails, archive, sourceClass);
```
Comment (jerrypeng) [inline pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionActioner.java]: This loads the NAR doesn't it?

File: pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionActioner.java [hunk 5]
```diff
@@ -414,50 +413,14 @@ private File getBuiltinArchive(FunctionDetails.Builder functionDetails) throws I
                 builder.setClassName(sinkClass);
                 functionDetails.setSink(builder);
 
-                fillSinkTypeClass(functionDetails, archive, sinkClass);
+                SinkConfigUtils.fillSinkTypeClass(functionDetails, archive, sinkClass);
```
Comment (jerrypeng) [inline pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionActioner.java]: Same comment as above

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java [hunk 6]
```diff
@@ -66,20 +66,16 @@
 import org.apache.pulsar.functions.sink.PulsarSinkDisable;
 import org.apache.pulsar.functions.source.PulsarSource;
 import org.apache.pulsar.functions.source.PulsarSourceConfig;
-import org.apache.pulsar.functions.utils.ComponentType;
-import org.apache.pulsar.functions.utils.Reflections;
-import org.apache.pulsar.functions.utils.FunctionCommon;
+import org.apache.pulsar.functions.utils.*;
```
Comment (jerrypeng) [inline pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java]: same here

**Comment by jerrypeng** [comment]
@srkukarni the PR in general looks fine since we really need to not load JARs in the worker/broker for security purposes.  I just have a few comments. However, the current validation logic is very complicated, hard to follow, and some methods are poorly named.  We should in the future, refactor the validation code.

Can we also have some unit/integration tests to test switching on and off the validation?

**Comment by srkukarni** [comment]
Closing this in favor of https://github.com/apache/pulsar/pull/4281
